File: blockchain-sdk/python/payments.py
"""
homeforai-blockchain — Cross-App Micropayment Protocol (Python)

Async implementation for Python services (Raven AI, OpenClinical AI, Hermes Edge).
Mirrors the TypeScript payments.ts implementation.
"""

from __future__ import annotations

import hashlib
import os
import re
from datetime import datetime, timezone
from typing import Literal, Optional

from .chains import CHAINS, ChainConfig


TargetApp = Literal["raven-ai", "openclinical-ai", "hermes-edge", "home-for-ai"]

APP_PAYMENT_RECEIVERS: dict[str, str] = {
    "raven-ai": "0xBadDB0b8Fe465E82B32D9b5c0E3571E214B9a5E1",
    "openclinical-ai": "0xC0fFee254729296a45a3885639AC7E10F9d54979",
    "hermes-edge": "0xDeAdBeEf000000000000000000000000dEAdbEEf",
    "home-for-ai": "0xABcDEf0123456789ABcDEf0123456789AbCdEf01",
}


async def pay_for_compute(
    amount: float,
    target_app: TargetApp,
    user_id: str,
    chain: str = "base",
    sender_address: Optional[str] = None,
) -> dict:
    """
    Creates a UserOperation to pay for compute in a target app.

    In production, signs and submits the UserOperation to a Bundler
    (e.g., Pimlico, Alchemy AA).

    Args:
        amount: Amount in USDC (e.g., 0.50 for $0.50)
        target_app: Target app to pay ('raven-ai', 'openclinical-ai', 'hermes-edge')
        user_id: User ID in the Home for AI ecosystem
        chain: Chain name (default: 'base')
        sender_address: User's smart account address (auto-generated if not provided)

    Returns:
        dict containing userOperation, mockTxHash, amountUsdc, etc.

    Example:
        result = await pay_for_compute(
            amount=0.50,
            target_app='raven-ai',
            user_id='user_123',
        )
        # Submit result['user_operation'] to bundler
        verified = await verify_payment(result['mock_tx_hash'])
    """
    chain_config = CHAINS.get(chain)
    if not chain_config:
        raise ValueError(f"Unknown chain: {chain}")

    receiver = APP_PAYMENT_RECEIVERS.get(target_app)
    if not receiver:
        raise ValueError(f"Unknown target app: {target_app}")

    amount_usdc = int(amount * 1_000_000)  # 6 decimals
    if not sender_address:
        sender_address = _generate_mock_smart_account(user_id)

    mock_tx_hash = "0x" + os.urandom(32).hex()

    user_op = {
        "sender": sender_address,
        "nonce": "0x0",
        "init_code": "0x",
        "call_data": _encode_usdc_transfer(chain_config.usdc, receiver, amount_usdc),
        "call_gas_limit": "0x493E0",
        "verification_gas_limit": "0x186A0",
        "pre_verification_gas": "0xC350",
        "max_fee_per_gas": "0x3B9ACA00",
        "max_priority_fee_per_gas": "0x3B9ACA00",
        "paymaster_and_data": "0x",
        "signature": "0x",
    }

    return {
        "user_operation": user_op,
        "mock_tx_hash": mock_tx_hash,
        "amount_usdc": f"{amount:.6f} USDC",
        "target_app": target_app,
        "chain": chain,
        "receiver": receiver,
        "estimated_gas_usd": "$0.001",
        "status": "pending",
        "created_at": datetime.now(timezone.utc).isoformat(),
    }


async def verify_payment(
    tx_hash: str,
    chain: str = "base",
) -> dict:
    """
    Verifies a payment by transaction hash.

    In production, queries the chain via RPC:
    1. eth_getTransactionReceipt(txHash)
    2. Parse UserOperationEvent log from EntryPoint
    3. Verify USDC Transfer event to the correct receiver

    Mock: returns verified=True for valid 0x-prefixed 66-char hashes.

    Args:
        tx_hash: Transaction hash (0x prefixed, 32 bytes)
        chain: Chain name (default: 'base')

    Returns:
        dict with 'verified' bool, 'tx_hash', 'chain', 'checked_at'

    Example:
        result = await verify_payment("0xabc123...", chain="base")
        if result['verified']:
            unlock_compute_for_user()
    """
    is_valid = bool(re.match(r"^0x[0-9a-fA-F]{64}$", tx_hash))
    return {
        "verified": is_valid,
        "tx_hash": tx_hash,
        "chain": chain,
        "checked_at": datetime.now(timezone.utc).isoformat(),
    }


async def get_balance(address: str, chain: str = "base") -> str:
    """
    Returns the USDC balance for an address.

    In production: calls eth_call on balanceOf(address) on the USDC contract.
    Mock: returns a deterministic demo balance.

    Args:
        address: EVM address (0x prefixed)
        chain: Chain name (default: 'base')

    Returns:
        Balance as a string (e.g., "100.00")
    """
    addr_num = int(address[2:8] or "0", 16)
    balance = 100 + (addr_num % 9000)
    return f"{balance:.2f}"


def build_payment_request_uri(
    target_app: TargetApp,
    amount: float,
    user_id: str,
    chain: str = "base",
) -> str:
    """
    Returns an EIP-681 URI for a cross-app payment request.
    Used to generate QR codes.
    """
    chain_config = CHAINS.get(chain)
    if not chain_config:
        raise ValueError(f"Unknown chain: {chain}")
    receiver = APP_PAYMENT_RECEIVERS.get(target_app, "")
    amount_usdc = int(amount * 1_000_000)
    return (
        f"ethereum:{chain_config.usdc}@{chain_config.chain_id}"
        f"/transfer?address={receiver}&uint256={amount_usdc}"
        f"&label={target_app}&memo={user_id}"
    )


# ─── Helpers ──────────────────────────────────────────────────────────────────

def _generate_mock_smart_account(user_id: str) -> str:
    h = hashlib.sha256(f"homeforai-account:{user_id}".encode()).hexdigest()
    return "0x" + h[:40]


def _encode_usdc_transfer(usdc_address: str, recipient: str, amount: int) -> str:
    """Encodes a USDC transfer call for the smart account."""
    # transfer(address,uint256): selector 0xa9059cbb
    selector = "a9059cbb"
    to_hex = recipient.replace("0x", "").lower().zfill(64)
    amount_hex = hex(amount)[2:].zfill(64)
    return "0x" + selector + to_hex + amount_hex

File: blockchain-sdk/src/payments.ts
/**
 * @homeforai/blockchain-sdk — Cross-App Micropayment Protocol
 *
 * Enables users to pay for services across the Home for AI ecosystem:
 * - Home for AI  → Raven AI (research compute)
 * - Home for AI  → OpenClinical AI (clinical AI inference)
 * - Home for AI  → Hermes Edge (premium model weights)
 *
 * Payment flow uses EIP-4337 Account Abstraction on Base chain (low gas, fast).
 * USDC is the default payment token on Base: 0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913
 */

import { CHAINS, ChainKey } from "./chains";
import { buildUserOperation, encodeErc20Transfer, encodeSmartAccountExecute, UserOperation } from "./eip4337";
import { randomBytes } from "node:crypto";

export type TargetApp = "raven-ai" | "openclinical-ai" | "hermes-edge" | "home-for-ai";

/** Receiving wallets for each app (replace with real treasury addresses in production) */
export const APP_PAYMENT_RECEIVERS: Record<TargetApp, string> = {
  "raven-ai": "0xBadDB0b8Fe465E82B32D9b5c0E3571E214B9a5E1",
  "openclinical-ai": "0xC0fFee254729296a45a3885639AC7E10F9d54979",
  "hermes-edge": "0xDeAdBeEf000000000000000000000000dEAdbEEf",
  "home-for-ai": "0xABcDEf0123456789ABcDEf0123456789AbCdEf01",
};

export interface PayForComputeParams {
  /** Amount in USDC (e.g., 0.50 for $0.50) */
  amount: number;
  /** Target app to pay */
  targetApp: TargetApp;
  /** User ID in the Home for AI ecosystem */
  userId: string;
  /** Chain to use (defaults to 'base') */
  chain?: ChainKey;
  /** User's smart account address (EIP-4337 account) */
  senderAddress?: string;
}

export interface PaymentResult {
  userOperation: UserOperation;
  /** Mock transaction hash — replace with real bundler submission */
  mockTxHash: string;
  amountUsdc: string;
  targetApp: TargetApp;
  chain: ChainKey;
  receiver: string;
  estimatedGasUsd: string;
  status: "pending" | "submitted";
  createdAt: string;
}

export interface VerifyPaymentResult {
  verified: boolean;
  txHash: string;
  chain: ChainKey;
  checkedAt: string;
}

/**
 * Creates a UserOperation to pay for compute in a target app.
 *
 * In production, this UserOperation would be signed with the user's smart
 * account key and submitted to a Bundler (e.g., Pimlico, Alchemy AA).
 *
 * @example
 * ```typescript
 * const result = await payForCompute({
 *   amount: 0.50,
 *   targetApp: 'raven-ai',
 *   userId: 'user_123',
 * });
 * // Submit result.userOperation to bundler
 * // Then poll verifyPayment(result.mockTxHash) until confirmed
 * ```
 */
export async function payForCompute(params: PayForComputeParams): Promise<PaymentResult> {
  const chain = params.chain ?? "base";
  const chainConfig = CHAINS[chain];
  const receiver = APP_PAYMENT_RECEIVERS[params.targetApp];

  // USDC has 6 decimals
  const amountUsdc = Math.round(params.amount * 1_000_000);
  const usdcAddress = chainConfig.usdc;

  // Encode USDC transfer: transfer(receiver, amount)
  const transferCallData = encodeErc20Transfer(receiver, BigInt(amountUsdc));

  // Encode smart account execute: execute(usdcAddress, 0, transferCallData)
  const callData = encodeSmartAccountExecute(usdcAddress, BigInt(0), transferCallData);

  const senderAddress = params.senderAddress ?? generateMockSmartAccountAddress(params.userId);

  const userOp = buildUserOperation({
    sender: senderAddress,
    callData,
    chain,
  });

  const mockTxHash = "0x" + randomBytes(32).toString("hex");

  return {
    userOperation: userOp,
    mockTxHash,
    amountUsdc: `${params.amount.toFixed(6)} USDC`,
    targetApp: params.targetApp,
    chain,
    receiver,
    estimatedGasUsd: "$0.001",  // Base is very cheap
    status: "pending",
    createdAt: new Date().toISOString(),
  };
}

/**
 * Verifies a payment by transaction hash.
 *
 * In production, this queries the chain via RPC:
 * 1. eth_getTransactionReceipt(txHash)
 * 2. Parse UserOperationEvent log from EntryPoint
 * 3. Verify USDC Transfer event to the correct receiver
 *
 * Mock implementation: returns true for any valid 0x-prefixed 66-char hash.
 */
export async function verifyPayment(
  txHash: string,
  chain: ChainKey = "base"
): Promise<VerifyPaymentResult> {
  const isValidHash = /^0x[0-9a-fA-F]{64}$/.test(txHash);

  return {
    verified: isValidHash,
    txHash,
    chain,
    checkedAt: new Date().toISOString(),
  };
}

/**
 * Returns the USDC balance for an address on a given chain.
 *
 * In production, calls eth_call on the USDC contract:
 * balanceOf(address) → uint256
 *
 * Mock implementation returns a realistic demo balance.
 */
export async function getBalance(
  address: string,
  chain: ChainKey = "base"
): Promise<string> {
  // Mock: returns a deterministic balance based on address
  const addrNum = parseInt(address.slice(2, 8), 16);
  const balance = (100 + (addrNum % 9000)).toFixed(2);
  return balance;
}

/**
 * Returns the full payment URL for a given target app and amount.
 * Used to generate QR codes for cross-app payment requests.
 */
export function buildPaymentRequestUri(
  targetApp: TargetApp,
  amount: number,
  userId: string,
  chain: ChainKey = "base"
): string {
  const receiver = APP_PAYMENT_RECEIVERS[targetApp];
  const chainConfig = CHAINS[chain];
  const amountUsdc = Math.round(amount * 1_000_000);
  // EIP-681 URI format for token transfer
  return `ethereum:${chainConfig.usdc}@${chainConfig.chainId}/transfer?address=${receiver}&uint256=${amountUsdc}&label=${targetApp}&memo=${userId}`;
}

// ─── Helpers ───────────────────────────────────────────────────────────────

function generateMockSmartAccountAddress(userId: string): string {
  const { createHash } = require("node:crypto");
  const hash = createHash("sha256").update("homeforai-account:" + userId).digest("hex");
  return "0x" + hash.slice(0, 40);
}
